### ISSUE TITLE
[Post-1.0] Automatic model-specific cognitive budget calibration

### ISSUE THREAD

**Issue by rinsakamo** [rinsakamo/relay-lm#1523]
# Purpose

Explore a post-1.0 capability that automatically calibrates RelayLM cognitive-budget settings when a user connects a different LLM model/provider configuration.

This is intentionally **post-1.0**. It must not expand the RelayLM 1.0 release scope or block #1388 Calibration / #1449 Release Readiness.

# Background

RelayLM v1 Calibration (#1388) is being reconstructed around a Fast Calibration / satisficing approach rather than exhaustive or one-token quality-boundary optimization.

The resulting machinery is potentially reusable beyond selecting one release default. It can serve as a bootstrap calibration engine for unknown or changed model deployments.

This revives an older product idea previously associated with RelayKV: automatically adapting RelayLM to the actual capacity and behavior of the connected model rather than requiring users to hand-tune context/memory budgets per model.

# Product idea

When a new or materially changed model deployment is detected:

1. fingerprint the effective model/provider configuration;
2. look for a compatible previously calibrated profile;
3. warm-start from an existing profile when compatibility is sufficiently strong;
4. otherwise run a bounded Fast Calibration procedure;
5. select a safe `good-enough` cognitive-budget profile;
6. persist the resulting profile with exact provenance;
7. reuse it until a recalibration trigger occurs.

Conceptually:

```text
model/provider configuration
        ↓
capability fingerprint
        ↓
compatible profile?
   ├─ yes → smoke validation → reuse/warm start
   └─ no  → Fast Calibration
                    ↓
          safe good-enough profile
                    ↓
             persist + reuse
```

# Candidate fingerprint dimensions

A reusable calibration identity likely needs more than a model display name. Consider at least:

- exact model artifact / revision where observable;
- quantization or serving artifact class where material;
- provider / adapter identity;
- tokenizer / serialized-input counter identity;
- effective context window;
- provider serialization/framing behavior;
- structured-output schema/capability set;
- State / Continuity proposal capabilities;
- decoding/seed capabilities;
- relevant RelayLM semantic/serialization contract versions.

The final compatibility key should be evidence-driven. Do not assume two deployments are interchangeable merely because they share a model family or context-window size.

# Calibration strategy

Reuse the post-reconstruction #1388 Fast Calibration approach where possible:

- deterministic/mechanical boundaries may use binary search, monotonic search, or semantic-owner change-point search;
- actual-model quality is treated as non-monotonic and should not use one-token binary search;
- coarse low/mid/high candidate screening should precede expensive replication;
- hard deterministic/safety constraints eliminate candidates early;
- successive halving / bounded coordinate refinement should allocate additional model calls only to survivors;
- stop once a safe, sufficiently good profile is found;
- prefer the smallest/simplest profile satisfying the acceptance constraints rather than searching for a theoretical optimum.

# Warm-start behavior

For a newly connected deployment, first attempt a cheap validation against a compatible existing profile.

Example:

```text
known compatible profile
        ↓
cheap deterministic + actual-model smoke
        ↓
PASS → reuse profile
FAIL → bounded Fast Calibration
```

Compatibility must fail closed when required model/provider/tokenizer/serialization information is unknown.

# Outputs

A successful automatic calibration should produce an auditable profile record equivalent to:

```text
profile id / version
applicable model/provider fingerprint
context window
reserved output capacity
State envelope/floor
Working Context envelope/floor
MEMORY envelope/floor
Event Evidence envelope/floor
legal degradation-policy identity
RelayLM calibration/evidence versions
source evidence IDs
known limitations
created/validated date
recalibration triggers
```

Exact shape remains future design work.

# Recalibration triggers

Potential triggers include:

- model artifact or quantization change;
- provider/adapter change;
- tokenizer/counter change;
- context-window change;
- prompt/serialization/framing change;
- structured-output schema change;
- relevant Context Compiler / Retrieval / Cognitive Budget semantic change;
- recurring pressure/overflow/product-quality failure;
- explicit operator request.

Not every semantic-only RelayLM change should force a full recalibration.

# Relationship to RelayKV

This Issue does **not** assume a specific future RelayKV architecture or require KV-cache implementation.

The historical connection is conceptual: the older product direction aimed to make model adaptation automatic. The current #1388 Fast Calibration work provides a concrete evidence-driven mechanism that could realize that idea after 1.0.

A future RelayKV or model-adaptation layer may consume these profiles, but ownership should be decided when post-1.0 architecture is designed.

# Non-goals

- expanding RelayLM 1.0 release scope;
- replacing #1388's current release-default calibration work;
- automatically changing semantic-owner rules;
- prompt tuning or fine-tuning the connected model;
- exhaustive hyperparameter optimization;
- assuming larger context windows always require proportionally larger budgets;
- selecting profiles by vendor/model-name folklore;
- building a KV-cache layer as part of this Issue without a separate owner decision.

# Design questions for post-1.0

- What exact fingerprint fields define profile compatibility?
- Which calibration checks can be fully deterministic and which require actual-model calls?
- What is the maximum acceptable bootstrap calibration cost/time?
- How should warm-start confidence be represented?
- Should profiles be local-only, portable, or shareable by capability class?
- How should evidence aging and automatic recalibration be surfaced to operators?
- Should calibration run automatically on first connection or require explicit opt-in?

# Acceptance direction

This Issue is ready for implementation planning after RelayLM 1.0 when:

1. #1388 Fast Calibration has a stable canonical workflow and at least one release profile;
2. the model/provider capability identity required for safe reuse is machine-readable;
3. a bounded calibration cost budget can be defined;
4. warm-start and fail-closed compatibility semantics are specified;
5. profile provenance and recalibration triggers have a stable storage contract.

Refs #1388 #1386 #1387 #1456
